resolver.go

```
@@ -250,13 +250,21 @@ func (r *Resolver) iterateParents(ctx context.Context, qname, qtype string, dept
 				return nil, ctx.Err()
 			case rrs := <-chanRRs:
 				for _, nrr := range nrrs {
-					if nrr.Name == qname {
+					if nrr.Name == qname && nrr.Type == "NS" {
 						rrs = append(rrs, nrr)
 					}
 				}
 				ctx := context.WithoutCancel(ctx)
 				cancel() // stop any other work here before recursing
-				return r.resolveCNAMEs(ctx, qname, qtype, rrs, depth)
+				res, err := r.resolveCNAMEs(ctx, qname, qtype, rrs, depth)
+
+				// for NS queries, return parent NS if we got
+				// no response
+				if res == nil && qtype == "NS" {
```

I apologize for missing this earlier.

What happens to the `err` above? It looks like it’s swallowed in this case. I know `resolveCNAMEs` doesn’t have a path that returns a non-nil error, but I feel like it might make sense to either check it or explicitly ignore it.

---

@kakwa can you get back to us here?

---

I've modified error handling to be more idiomatic.